### ISSUE TITLE
Automate the sim_deployment.md steps

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#616]
We should be able to provide high level directives that automate the entirety of the sim_deployment steps.

**Comment by ariel-phet** [comment]
I imagine this comment will be relevant 

https://github.com/phetsims/chipper/issues/560#issuecomment-342227691

**Comment by jonathanolson** [comment]
I'm concerned about the "manual test" that happens to the built version in the middle (currently the "Test the built version locally" step).

Can we avoid this, and say that the developer can test before kicking off the deploy?

**Comment by samreid** [comment]
> Can we avoid this, and say that the developer can test before kicking off the deploy?

Yes, this was important when the build tools were in their infancy, but now a pre-test and a post-deploy test would be sufficient.

**Comment by jonathanolson** [comment]
Sounds good, thanks.

**Comment by jonathanolson** [comment]
Added perennial `grunt create-release --repo={{REPO}} --branch={{BRANCH}}` to handle creating a new release branch (and properly pushing master and branch version updates).

**Comment by jonathanolson** [comment]
Initial implementation for dev/rc/production deployment is done. I've tested dev deployment somewhat with both brands (I'll want to test all failure modes too).

I'll probably need to look at build-server (possibly with @mattpen) to see what the consequences of changing the phet-brand "version" to include `-phet` in the build-server request, and also how to handle the other naming changes in https://github.com/phetsims/chipper/issues/560#issuecomment-342227691.

**Comment by jonathanolson** [comment]
(rc and dev in perennial are totally untested. please for the love of god don't run them).

**Comment by pixelzoom** [comment]
😈

**Comment by jonathanolson** [comment]
RC deploy looks like it's working (https://www.colorado.edu/physics/phet/dev/html/chains/1.6.0-rc.1-phet/chains_en-phet.html, note the new naming scheme)

**Comment by jonathanolson** [comment]
@mattpen, `sudo journalctl -fu build-server` on ox-dev says `sudo: journalctl: command not found`. Is there a way I can look at the logs?

**Comment by jonathanolson** [comment]
phet-io deploy kinda looks like it worked (https://www.colorado.edu/physics/phet/dev/html/chains/1.6.0-rc.2-phetio/wrappers/index), but I can't test.

Why we have to deploy rc.1 and rc.2 is on my list of things to discuss on Thursday. I'd like to have dev/rc deploys for phet-io-enabled sims to by default deploy both with the same version (thus same shas).

**Comment by jonathanolson** [comment]
"production" (to ox-dev) deploy failed after sending build request on the local side (something related to updating third-party-licenses).

I'll also need an improved way of detecting (or asking) if it was a first production deploy, for updating the readme (instead of checking if it is 1.0.0).

**Comment by jonathanolson** [comment]
`Fatal error: Unable to delete "downloaded-sims" file (ENOTEMPTY: directory not empty, rmdir 'C:\Users\olson\phet\git\chipper\downloaded-sims').`

It's likely this was caused by me, but @samreid does that error message look familiar?

**Comment by jonathanolson** [comment]
@mattpen, 1.6.0 of chains never showed up on dev or phet-dev. I'll want to check the build logs to see what went wrong. It also doesn't seem to be sending emails when sending to phet-dev (could be my fault).

**Comment by samreid** [comment]
> Fatal error: Unable to delete "downloaded-sims" file (ENOTEMPTY: directory not empty, rmdir 'C:\Users\olson\phet\git\chipper\downloaded-sims').

The code looks like this:

```js
  if ( grunt.file.exists( 'downloaded-sims' ) ) {
    grunt.file.delete( 'downloaded-sims' );
  }
```

According to the docs at https://gruntjs.com/api/grunt.file#grunt.file.delete

> Delete the specified filepath. Will delete files and folders recursively.

> Will not delete the current working directory or files outside the current working directory unless the --force command-line option is specified.

I wonder if prior usage was inside the current working directory, but chipper 2.0 was trying to work in a different directory?

**Comment by jonathanolson** [comment]
> I wonder if prior usage was inside the current working directory, but chipper 2.0 was trying to work in a different directory?

This seems to be failing in master also?

**Comment by jonathanolson** [comment]
It also seems to be legitimately empty.

**Comment by samreid** [comment]
Not sure, possible that it was never properly tested before (running with an existing downloaded-sims directory.  Or maybe our node updates have caused this?  (Not sure what node filesystem API grunt is using).

**Comment by jonathanolson** [comment]
Question: Should dev deploys double-build the sim? (Build it once before changing the version and prompting the user, to make sure it builds correctly without failures? Then build it again after committing/pushing the version?). It is fairly easy to just detect the failure in the second step build-side and revert any version changes, and it would probably save developer time overall to do that.

**Comment by jonathanolson** [comment]
As noted in https://github.com/phetsims/perennial/issues/1, we can't currently run perennial grunt commands from sims. Since the new automated deployment steps currently exist in perennial, should this functionality be added?

Notably, we'd need to make sure that our perennial gruntfile doesn't interfere with ANY PAST chipper gruntfile, which seems trickly.

Alternatively, we could create `bin/perennial` under each sim, which would allow running perennial grunt commands.

**Comment by samreid** [comment]
> Question: Should dev deploys double-build the sim?

I'm fine either way.  Rolling back on fail would introduce spurious commits, but hopefully it would be rare.

**Comment by mattpen** [comment]
> sudo journalctl -fu build-server on ox-dev says sudo: journalctl: command not found. Is there a way I can look at the logs?

Jon and I discussed this on slack.  The testing server is now phet-server-dev.int.colorado.edu.  I'll add this to the documentation in build-server.js.

> 1.6.0 of chains never showed up on dev or phet-dev.

@jonathanolson - I'm guessing `dev = spot` and `phet-dev=(ox-dev or phet-server-dev)`?  

Regarding SPOT:  The build-server assumes that the client is doing the spot deploy during a production build.  The only time versions are deployed to spot from the build-server (as of now) is in an RC deploy, so we should take a look at the local log output from the production build to determine why it didn't deploy to spot.  Should we consider migrating this task to perennial? 

Regarding OX-DEV: The production version was deployed to ox-dev/phet-server-dev, you can see it here: https://ox-dev.colorado.edu/sims/html/chains/1.6.0/. 
- It looks like there are a some files we don't want in production on the phet brand website, i.e. https://ox-dev.colorado.edu/sims/html/chains/1.6.0/chains_en-phetio.html.
- The file names for the phet branded sims changed, which means `latest` links will change, i.e. we now see https://ox-dev.colorado.edu/sims/html/chains/latest/chains_en-phet.html and https://ox-dev.colorado.edu/sims/html/chains/latest/chains_en.html is broken.

**Comment by jonathanolson** [comment]
> The file names for the phet branded sims changed, which means latest links will change.

> "-phet" is stripped out of directory and file names when production versions are deployed to the production server. (AP, KP: We will not change the names on the production server, too problematic.)

(as noted in https://github.com/phetsims/chipper/issues/560). So the build-server should handle the renaming when moving things to production.

**Comment by mattpen** [comment]
@jonathanolson - Do we want all the files that are now created by a production deploy copied to the public document root?

Here is the directory for chains from before and after the chages:
```
[mape5853@phet-server-dev 1.6.0]$ ls -l
total 14992
-rw-rw-r--.  1 tomcat phet    9158 Nov 28 21:41 chains-128.png
-rw-rw-r--.  1 tomcat phet   54546 Nov 28 21:41 chains-600.png
-rw-rw-r--.  1 tomcat phet 1072409 Nov 28 21:41 chains_ab.html
-rw-rw-r--.  1 tomcat phet 1168045 Nov 28 21:41 chains_ab-phet.html
-rw-rw-r--.  1 tomcat phet 1168055 Nov 28 21:41 chains_da-phet.html
-rw-rw-r--.  1 tomcat phet 1167986 Nov 28 21:41 chains_en_CA-phet.html
-rw-rw-r--.  1 tomcat phet     895 Nov 28 21:41 chains_en-iframe.html
-rw-rw-r--.  1 tomcat phet 1167970 Nov 28 21:41 chains_en-phet.html
-rw-rw-r--.  1 tomcat phet 1231560 Nov 28 21:41 chains_en-phetio.html
-rw-rw-r--.  1 tomcat phet 1168108 Nov 28 21:41 chains_es_MX-phet.html
-rw-rw-r--.  1 tomcat phet 1168227 Nov 28 21:41 chains_ga-phet.html
-rw-rw-r--.  1 tomcat phet 1169279 Nov 28 21:41 chains_kn-phet.html
-rw-rw-r--.  1 tomcat phet 1167969 Nov 28 21:41 chains_lk-phet.html
-rw-rw-r--.  1 tomcat phet 1168071 Nov 28 21:41 chains_nl-phet.html
-rw-rw-r--.  1 tomcat phet 1167998 Nov 28 21:41 chains_ro-phet.html
-rw-rw-r--.  1 tomcat phet   67528 Nov 28 21:41 chains-twitter-card.png
-rw-rw-r--.  1 tomcat phet     208 Nov 28 21:41 chains.xml
-rw-rw-r--.  1 tomcat phet 1167898 Nov 28 21:41 chains_zh_CN-phet.html
drwxrwsr-x.  3 tomcat phet    4096 Nov 28 21:41 contrib
-rw-rw-r--.  1 tomcat phet    2268 Nov 28 21:41 dependencies.json
drwxrwsr-x.  4 tomcat phet      63 Nov 28 21:41 docs
drwxrwsr-x.  2 tomcat phet      23 Nov 28 21:41 lib
drwxrwsr-x. 16 tomcat phet    4096 Nov 28 21:41 wrappers
[mape5853@phet-server-dev 1.6.0]$ cd ../1.2.8/
[mape5853@phet-server-dev 1.2.8]$ ls -l
total 9448
-rw-rw-r--. 1 tomcat phet 1072409 Nov 28 11:50 chains_ab.html
-rw-rw-r--. 1 tomcat phet 1072424 Nov 16 13:52 chains_da.html
-rw-rw-r--. 1 tomcat phet 1072350 Nov 16 13:52 chains_en_CA.html
-rw-rw-r--. 1 tomcat phet 1072334 Nov 16 13:52 chains_en.html
-rw-rw-r--. 1 tomcat phet     895 Nov 16 13:52 chains_en-iframe.html
-rw-rw-r--. 1 tomcat phet 1072475 Nov 16 13:52 chains_es_MX.html
-rw-rw-r--. 1 tomcat phet 1073643 Nov 16 13:52 chains_kn.html
-rw-rw-r--. 1 tomcat phet 1072333 Nov 16 13:52 chains_lk.html
-rw-rw-r--. 1 tomcat phet 1072362 Nov 16 13:52 chains_ro.html
-rw-rw-r--. 1 tomcat phet     936 Nov 28 11:50 chains.xml
-rw-rw-r--. 1 tomcat phet 1072274 Nov 16 13:52 chains_zh_CN.html
-rw-rw-r--. 1 tomcat phet    1611 Nov 28 11:50 dependencies.json
```

It looks like we should only copy the following:
-  `{{anything}}.png`
- `{{sim-name}}_{{locale}}.html` (after renaming from `{{sim-name}}_{{locale}}-phet.html`
- `{{sim-name}}.xml`
- `{{sim-name}}_en-iframe.html`
- `dependencies.json`

Does this sound right to you?  The alternative would be to exclude the following files:
- `contrib/`
- `docs/`
- `wrappers/`
- `lib/`
- `{{sim-name}}_{{locale}}-phetio.html`

**Comment by mattpen** [comment]
The previous comment may depend on decisions in #630.

**Comment by zepumph** [comment]
I think we are trying to push for only _all HTML files for phet-io built files. To be confirmed tomorrow at meetings though.

**Comment by jonathanolson** [comment]
> @jonathanolson - Do we want all the files that are now created by a production deploy copied to the public document root?

I'm honestly a little confused at how a phetio deploy put files into the phet document root. Isn't that not supposed to happen, so this won't be an issue at all?

**Comment by mattpen** [comment]
It looks odd to me too but I had assumed that changes for the ["build both brands in the same deploy"](https://github.com/phetsims/chipper/issues/567) had been implemented already.

**Comment by jonathanolson** [comment]
Nope. I think that may still fire off two separate build-server requests.

**Comment by samreid** [comment]
> So the build-server should handle the renaming when moving things to production.

In my opinion, things will be simpler if we use the production name on the dev server.  The hassle of dealing with renaming and understanding the different names will be worse than just dealing with production names in the first place.

**Comment by samreid** [comment]
> I'm honestly a little confused at how a phetio deploy put files into the phet document root. 

I'm not sure, but @zepumph may know.

**Comment by zepumph** [comment]
I'm not sure I understand fully, I wasn't aware that a phetio deploy would touch the phet.colorado.edu. Am I understanding this right?

**Comment by mattpen** [comment]
@zepumph - The latest RC and production deploys for chains for phet brand have included the phetio artifacts.  See https://ox-dev.colorado.edu/sims/html/chains/1.6.0/.  This was deployed as version `1.6.0-phet`.

**Comment by zepumph** [comment]
Is this because the chipper 2.0 build now builds with both (all) brands, but to deploy just copied all artifacts to the phet site?  It seems like it would be nice to separate brands out to different folders if we are always building all brands (at least by default). Otherwise it could be so confusing to see all of that intermingled, taking intimate knowledge of all brands to be able to split off different files.

**Comment by jonathanolson** [comment]
I'll look into what's happening, it may be that the "clean" step was missed.

**Comment by jonathanolson** [comment]
@mattpen, I'll be ready to collaborate on build-server changes so that it supports chipper 2.0 build structures (see #560).

I updated the dev deploy script (e.g. https://www.colorado.edu/physics/phet/dev/html/chains/1.7.0-dev.9/) for this, and I'll add the ability to add one-offs.

**Comment by jonathanolson** [comment]
@zepumph, as part of this, I've moved the sonification wrapper build/deploy to the new system. I deployed https://www.colorado.edu/physics/phet/dev/html/phet-io-wrapper-sonification/1.0.0-dev.41/, but it may have a few issues (not loading in a few spot checks). I had to push a lint fix to the repo before competing testing, so I'm not sure what condition it is in.

This is testable by checking out (and pulling) chipper branch 2.0, and in perennial (after npm prune/update in perennial):
```sh
grunt wrapper --repo=phet-io-wrapper-sonification
```
(don't worry, the wrapper repo and chipper will get npm updated as part of the process)

Can you see if something I did is incorrect?

Main two files:

- https://github.com/phetsims/chipper/blob/2.0/js/grunt/phet-io/buildWrapper.js (builds the wrapper into the build/ directory)
- https://github.com/phetsims/perennial/blob/master/js/grunt/wrapper.js (deploys the wrapper - some temporary copied code shared with dev deployment that will be factored out)

**Comment by jonathanolson** [comment]
FYI a simple `grunt` in the wrapper directory should build it (hopefully that makes the build easier to test).

**Comment by zepumph** [comment]
The above comments were responded to in phetsims/phet-io#1099.

**Comment by jonathanolson** [comment]
Tagged because I'd like to know preferences on final filenames. Should we have:

- phet-info/deployment-info/chipper-1.0.md (currently sim_deployment.md)
- phet-info/deployment-info/chipper-2.0.md

or some other form of organization? I'd still like to keep the 1.0 checklists around, since it's helpful for maintenance releases and reference.

**Comment by samreid** [comment]
My preference would be to overwrite phet-info/deployment-info/sim_deployment.md with the 2.0 instructions.  If you want the old instructions, you can see them in the git history.

**Comment by pixelzoom** [comment]
How handy does the 1.0 checklist need to be? If it's going to be needed frequently, then I don't like the idea of having to dig it out of git history.

**Comment by samreid** [comment]
If you're going to be doing a deploy from chipper 1.0, then you probably have the 1.0 checklist checked out.

**Comment by pixelzoom** [comment]
How so? sim_deployment.md (the check list) lives in phet-info, not a sim that typically in dependencies.json.

**Comment by samreid** [comment]
Ah, my mistake.

**Comment by jonathanolson** [comment]
> How handy does the 1.0 checklist need to be? If it's going to be needed frequently, then I don't like the idea of having to dig it out of git history.

Depends on how many maintenance releases we'd end up doing on sim branches that have already been published. Having documentation available in master (until we have no published chipper <2.0 sims left that would need maintenance) just sounded easier to me.

Once all sims are chipper 2.0+, we'll want to remove legacy support from things like the build-server, perennial/etc., and we can clean up the documentation then too.

**Comment by pixelzoom** [comment]
+1 for having both sets of doc available separately in master, and cleaning up in the future (as chipper 1.0 needs taper off).  I don't care what they're named, as long as it's clear which is which.

**Comment by jonathanolson** [comment]
I'll also want to update sim_deployment.md references in bulk-maintenance-release-process.md, chipper-2.0.md and deployment-info/README.md (note to my future self).

I'm not aware of Google Doc references, but there probably are some?

**Comment by samreid** [comment]
> +1 for having both sets of doc available separately in master, and cleaning up in the future (as chipper 1.0 needs taper off). I don't care what they're named, as long as it's clear which is which.

Sounds good to me.

**Comment by pixelzoom** [comment]
1/18/18 dev meeting:  

Both docs will be available in master, named chipper-1.0.md and chipper-2.0.md. @jonathanolson will make the changes, and search for potential Google Doc reference changes.

**Comment by jonathanolson** [comment]
Applied documentation updates. Searched google docs, but it only brought up status/developer meeting notes. I don't know of any other references.

Closing, as I believe this issue is now complete.
